modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/BinaryInsertCoercionTest.java

```diff
@@ -0,0 +1,159 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.planner.datatypes;
+
+import static org.apache.calcite.sql.type.SqlTypeName.BINARY_TYPES;
+import static org.apache.ignite.internal.lang.IgniteStringFormatter.format;
+import static org.apache.ignite.internal.sql.engine.util.TypeUtils.native2relationalType;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import java.util.List;
+import java.util.stream.Stream;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.BinaryPair;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.TypePair;
+import org.apache.ignite.internal.sql.engine.planner.datatypes.utils.Types;
+import org.apache.ignite.internal.sql.engine.rel.IgniteKeyValueModify;
+import org.apache.ignite.internal.sql.engine.rel.IgniteRel;
+import org.apache.ignite.internal.sql.engine.schema.IgniteSchema;
+import org.apache.ignite.internal.sql.engine.type.IgniteTypeFactory;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.type.NativeType;
+import org.apache.ignite.internal.type.VarlenNativeType;
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * A set of tests to verify behavior of type coercion for INSERT operations, when values belongs to the BINARY type family.
+ *
+ * <p>This tests aim to help to understand in which cases implicit cast will be added to which values.
+ */
+public class BinaryInsertCoercionTest extends BaseTypeCoercionTest {
+    @ParameterizedTest
+    @MethodSource("args")
+    public void insert(
+            TypePair pair,
+            Matcher<RexNode> operandMatcher
+    ) throws Exception {
+        IgniteSchema schema = createSchemaWithSingleColumnTable(pair.first());
+
+        int length = ((VarlenNativeType) pair.second()).length();
+
+        String byteVal = "x'" + "01".repeat(length) + "'";
+
+        assertPlan("INSERT INTO T VALUES(" + byteVal + ")", schema, keyValOperandMatcher(operandMatcher)::matches, List.of());
+    }
+
+    @ParameterizedTest
+    @MethodSource("argsDyn")
+    public void insertDynamicParameters(
+            TypePair pair,
+            Matcher<RexNode> operandMatcher
+    ) throws Exception {
+        IgniteSchema schema = createSchemaWithSingleColumnTable(pair.first());
+
+        int length = ((VarlenNativeType) pair.second()).length();
+
+        byte[] byteVal = new byte[length];
+
+        assertPlan("INSERT INTO T VALUES(?)", schema, keyValOperandMatcher(operandMatcher)::matches, List.of(byteVal));
+    }
+
```

Comment:
please add test to make sure we haven't forgotten any type pair. See `org.apache.ignite.internal.sql.engine.planner.datatypes.NumericInsertSourcesCoercionTest#insertArgsIncludesAllTypePairs` for example

Reply:
done